p2panda-stream/src/ordering/partial/operations.rs

```diff
@@ -0,0 +1,376 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::{collections::HashMap, marker::PhantomData};
+
+use p2panda_core::{Extensions, Hash, Operation};
+use p2panda_store::{LogStore, OperationStore};
+use thiserror::Error;
+
+use crate::ordering::{PartialOrder as InnerPartialOrder, PartialOrderError, PartialOrderStore};
+
+/// Struct for processing p2panda operations into a partial order based on dependencies expressed
+/// in their `previous` field.
+///
+/// This struct is a thin wrapper around ordering::PartialOrder struct which takes care of sorting
+/// the operation DAG into a partial order. Here we have the addition of a `LogStore` and
+/// `OperationStore` implementation (traits from `p2panda-store`) and an operation cache which
+/// helps us avoid unnecessary calls to the database.
+pub struct PartialOrder<L, E, OS, POS> {
+    /// A store containing p2panda operations.
+    operation_store: OS,
+
+    /// The inner PartialOrder struct which sorts the operation DAG into a partial order.
+    inner: InnerPartialOrder<Hash, POS>,
+
+    /// In memory cache of operations we have processed, before retrieving "ready" operations from
+    /// the store they will be taken from here if present. The store can be cleared at any time
+    /// in order to reduce memory use.
+    operation_cache: HashMap<Hash, Operation<E>>,
+    _phantom: PhantomData<(L, E)>,
+}
+
+impl<L, E, OS, POS> PartialOrder<L, E, OS, POS>
+where
+    OS: OperationStore<L, E> + LogStore<L, E>,
+    POS: PartialOrderStore<Hash>,
+    E: Extensions,
+{
+    /// Construct a new dependency checker from an operation store and dependency store.
+    pub fn new(operation_store: OS, partial_order_store: POS) -> Self {
+        let inner_dependency_checker = InnerPartialOrder::new(partial_order_store);
+        PartialOrder {
+            operation_store,
+            inner: inner_dependency_checker,
+            operation_cache: Default::default(),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Process a single operation.
+    pub async fn process(
+        &mut self,
+        operation: Operation<E>,
+    ) -> Result<(), OperationDependencyCheckerError> {
+        let hash = operation.hash;
+        let previous = operation.header.previous.clone();
+        self.operation_cache.insert(operation.hash, operation);
+        self.inner.process(hash, previous).await?;
+        Ok(())
+    }
+
+    /// Take the next ready operation from the queue.
+    pub async fn next(&mut self) -> Result<Option<Operation<E>>, OperationDependencyCheckerError> {
+        let Some(hash) = self.inner.next() else {
+            return Ok(None);
+        };
+
+        if let Some(operation) = self.operation_cache.remove(&hash) {
+            return Ok(Some(operation));
+        }
+
+        if let Some((header, body)) = self
+            .operation_store
```

Comment:
Interesting that this assumes that the operations are already stored, maybe this needs to be mentioned somewhere?

If we ingest operations _after_ the ordering we might need to change that in the future.